Support for Java8 parameters

cbxp opened the pull request:
Instead of slow additional enhancement in `LocalvariablesNamesEnhancer`.

Java 7 is still supported, but when running on Java 8, there will be less enhancement because Java 8 standard `Method.getParameters()` will be used to get parameter names.

Later, if we drop support for Java 7 in some next release, we can just delete `LocalvariablesNamesEnhancerJava7` class.

`LocalvariablesNamesEnhancerJava7` is just a copy of old `LocalvariablesNamesEnhancer`. Plain `LocalvariablesNamesEnhancer` now has the signature enhancement code removed.

asolntsev commented:
@xael-fry I believe this is a good PR. It solves the issue https://github.com/playframework/play1/issues/1043

Please review it.

xael-fry commented on framework/src/play/utils/Java.java:
```diff
@@ -194,10 +194,15 @@ public static Object invokeStatic(Method method, Object[] args) throws Exception
+            // No import, so Java 7 will not break trying to load this class
+            java.lang.reflect.Parameter[] parameters = method.getParameters();
+            String[] names = new String[parameters.length];
+            for (int i = 0; i < parameters.length; i++) names[i] = parameters[i].getName();
```
Don't use inline delclaration for ```for```

cbxp commented:
This PR actually works with Java 7.
If you mean that in 1.5.x we can drop Java 7 support, then this PR can be
simplified.

On Tue, 10 Jan 2017, 08:03 Alexandre Chatiron, <notifications@github.com>
wrote:

> *@xael-fry* requested changes on this pull request.
>
> Please squash your change in one commit.
> We will create a new branch 1.5.x and to merge this PR in.
> Seems like the test is also failing on 1.4.x branch. I will check it
> ------------------------------
>
> In framework/src/play/utils/Java.java
> <https://github.com/playframework/play1/pull/1084#pullrequestreview-15853761>
> :
>
> > @@ -194,10 +194,15 @@ public static Object invokeStatic(Method method, Object[] args) throws Exception
>       * Retrieve parameter names of a method
>       */
>      public static String[] parameterNames(Method method) throws Exception {
> -        try {
> -            return (String[]) method.getDeclaringClass().getDeclaredField("$" + method.getName() + LocalVariablesNamesTracer.computeMethodHash(method.getParameterTypes())).get(null);
> -        } catch (Exception e) {
> -            throw new UnexpectedException("Cannot read parameter names for " + method, e);
> +        if (Play.classes.java8) {
> +            // No import, so Java 7 will not break trying to load this class
> +            java.lang.reflect.Parameter[] parameters = method.getParameters();
> +            String[] names = new String[parameters.length];
> +            for (int i = 0; i < parameters.length; i++) names[i] = parameters[i].getName();
>
> Don't use inline delclaration for for
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/playframework/play1/pull/1084#pullrequestreview-15853761>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABmoKbUkIJatvulDKfZDMZvlkBmYZhe1ks5rQx8ugaJpZM4LeYgX>
> .
>

xael-fry commented:
Sorry my bad. so we will merge in 1.4.x after remarks fixes

cbxp commented:
I have fixed the for loop as was requested, and also squashed some commits, but not all.

Squashing many changes with their own reasoning (commit messages) into a single big commit usually a not good idea, as it is quite hard or impossible to understand later the reasoning behind particular changes.

Hopefully these 3 remaining commits are ok.

cbxp commented:
The build seems to be failing in 1.4.x already...

dejan2609 commented:
> The build seems to be failing in 1.4.x already...

Indeed, see here for more details (solution proposed by @asolntsev):
- https://github.com/playframework/play1/pull/1086
- [Tests are failing (for branch 1.4.x) both locally and on GitHub/Travis](https://play.lighthouseapp.com/projects/57987/tickets/2076)

xael-fry commented on .travis.yml:
```diff
@@ -1,7 +1,7 @@
 sudo: required
 language: java
 jdk:
-  - openjdk7 
```
Could you keep ```openjdk7``` to be sure 1.4.x still compile with 1.7?

cbxp replied:
Then we cannot build play... It would be nice to build with Java 8, but run with Java 7, but I don't know how to do that with travis

asolntsev commented:
@xael-fry @angryziber @cbxp it's possible

```yml
jdk:
  - oraclejdk8
  - openjdk7
addons:
  apt:
    packages:
      - oracle-java8-installer

script:
  - jdk_switcher use oraclejdk8
  - # do stuff with Java 8
  - jdk_switcher use oraclejdk7
  - # do stuff with Java 7
```

See for more details: https://docs.travis-ci.com/user/languages/java/

xael-fry commented:
But after thinking about it, it we cannot compile it with JDK 7 we will not merge it in 1.4.x. but in a new branch 1.5.x.

angryziber commented:
Dropped Java 7 support - now this can be merged to the newly created master branch for the upcoming 1.5.x release series

asolntsev commented:
merged to `master` branch

xael-fry commented:
All travis tests of this PR fails, reset master

angryziber commented:
Sorry, originally it specified oraclejdk8, which was correct for Travis.
Later it was changed to openjdk8, which is not supported.
Andrei, can you please change it to oraclejdk8 in .travis.yml?

Anton

On Mon, Feb 6, 2017 at 3:01 AM Alexandre Chatiron <notifications@github.com>
wrote:

> All travis tests of this PR fails, reset master
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/playframework/play1/pull/1084#issuecomment-277564920>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA11YFqVFuNoX9dTMWgkuRVThuN4C2Lnks5rZnDhgaJpZM4LeYgX>
> .
>
-- 

Anton

angryziber commented:
@xael-fry can you please reopen this PR because it was removed from master?

xael-fry commented:
will be handle in #1099

angryziber commented:
Do you want me to create a new pull request or you don't want to use my
commits at all?

On Tue, 7 Feb 2017, 07:07 Alexandre Chatiron, <notifications@github.com>
wrote:

> will be handle in #1099 <https://github.com/playframework/play1/pull/1099>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/playframework/play1/pull/1084#issuecomment-277903044>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA11YAeNulVloBbatjRvgGoWHXWVT1PVks5rZ_wjgaJpZM4LeYgX>
> .
>
-- 

Anton

xael-fry commented:
@angryziber just look at #1099, I rebase your PR and merged it in master

xael-fry commented:
@angryziber just look at #1099, I rebase your PR and merged it in master
